Add NrgKick Gen2

wobu opened the pull request:
Fix https://github.com/evcc-io/evcc/issues/14786

I added an initial implementatin for the NRGKick Gen2 Charter and his new JSON API

https://www.nrgkick.com/wp-content/uploads/2024/07/local_api_docu_simulate.html

I own a NRG Kick Gen 2 and my JSON Api got activated just today, so i just had to implement the charger right away.
This implementation is untested at the moment, but i wanted to share it nevertheless. It is my first pull request to this community and feedback is welcome. 

When i have a running evcc setup i will test this implementation. Unfortunatley i am also missing a BEV to test with :P

wobu committed:
Added types and initial implementation for NrgKick Gen2 charger json API

wobu committed:
added nrgkick gen2 template and added to  README.md

wobu committed:
fixed nrgkick gen2 types

wobu commented:
evcc could successfully connect to my nrgkick gen2 charger. Values could be read. Unfortunately i don't know how to test any value changes, like current

wobu committed:
fixed type for nrgkick gen2 Control.CurrentSet

wobu commented:
![grafik](https://github.com/user-attachments/assets/72a0958d-63f5-4595-9406-520c0667e226)

i don't know if there is still something wrong with the shown voltages

Ahhpaddy commented:
Hello I can test, I have 2 Nrgkick 's.
But I don't know how to setup my evcc.yaml
Can you maybe explain how to do that?

I'm on v0.128.2

wobu committed:
fixed watt hours conversion for nrgkick gen2

wobu committed:
fixed missing colons in nrgkick gen2 types

wobu commented:
voltages are now working too:

![grafik](https://github.com/user-attachments/assets/f6015d4c-750c-4912-b5a4-f2712dfd11fa)

wobu committed:
fixed changing current for nrgkick gen2

wobu commented:
i tested via command line the switching of the current. Working now.

> PS C:\code\evcc> .\evcc.exe charger
[main  ] INFO 2024/07/20 11:20:09 evcc 0.0.0
[main  ] INFO 2024/07/20 11:20:09 using config file: C:\code\evcc\evcc.yaml
Power:          1W
Energy:         185.0kWh
Current L1..L3: 0.024A 0.024A 0.024A
Voltage L1..L3: 236V 236V 236V
Charge status:  A
Enabled:        false
Charged:        0.0kWh
Max Current:    15.0A
Phases:         3
PS C:\code\evcc> .\evcc.exe charger -i 12
[main  ] INFO 2024/07/20 11:21:33 evcc 0.0.0
[main  ] INFO 2024/07/20 11:21:33 using config file: C:\code\evcc\evcc.yaml
PS C:\code\evcc> .\evcc.exe charger
[main  ] INFO 2024/07/20 11:21:36 evcc 0.0.0
[main  ] INFO 2024/07/20 11:21:36 using config file: C:\code\evcc\evcc.yaml
Power:          1W
Energy:         185.0kWh
Current L1..L3: 0.025A 0.025A 0.025A
Voltage L1..L3: 236V 236V 236V
Charge status:  A
Enabled:        false
Charged:        0.0kWh
Max Current:    12.0A
Phases:         3
PS C:\code\evcc> .\evcc.exe charger -i 16
[main  ] INFO 2024/07/20 11:21:38 evcc 0.0.0
[main  ] INFO 2024/07/20 11:21:38 using config file: C:\code\evcc\evcc.yaml
PS C:\code\evcc> .\evcc.exe charger
[main  ] INFO 2024/07/20 11:21:41 evcc 0.0.0
[main  ] INFO 2024/07/20 11:21:41 using config file: C:\code\evcc\evcc.yaml
Power:          1W
Energy:         185.0kWh
Current L1..L3: 0.025A 0.025A 0.025A
Voltage L1..L3: 236V 236V 236V
Charge status:  A
Enabled:        false
Charged:        0.0kWh
Max Current:    16.0A
Phases:         3
PS C:\code\evcc> .\evcc.exe charger -i 18
[main  ] INFO 2024/07/20 11:21:43 evcc 0.0.0
[main  ] INFO 2024/07/20 11:21:43 using config file: C:\code\evcc\evcc.yaml
PS C:\code\evcc> .\evcc.exe charger
[main  ] INFO 2024/07/20 11:21:45 evcc 0.0.0
[main  ] INFO 2024/07/20 11:21:45 using config file: C:\code\evcc\evcc.yaml
Power:          1W
Energy:         185.0kWh
Current L1..L3: 0.027A 0.027A 0.027A
Voltage L1..L3: 236V 236V 236V
Charge status:  A
Enabled:        false
Charged:        0.0kWh
Max Current:    16.0A
Phases:         3
PS C:\code\evcc>

wobu committed:
fixed GetPhases for nrgkick gen2

wobu committed:
updated status and error code handling for NRGKick gen2

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+
+	var res gen2.Control
+
+	if err := nrg.DoJSON(req, &res); err != nil {
```
Why not use `GetJSON` similar to https://github.com/evcc-io/evcc/pull/14938/files#diff-e66d191361cf8ba2a343492ad6d253ad448eddd02ab4c4630df3cf3478c4084dR69?

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+// Enable implements the api.Charger interface
+func (nrg *NRGKickGen2) Enable(enable bool) error {
+	res, err := nrg.controlG.Get()
+
```
```suggestion
```

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+		return err
+	}
+
+	if enable {
```
res.ChargePause = cast.ToInt(!enable)

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+	}
+
+	err = nrg.updateControl(res, false)
+
```
```suggestion
```

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+	}
+
+	res, err := nrg.controlG.Get()
+
```
```suggestion
```

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+
+func (nrg *NRGKickGen2) GetMaxCurrent() (float64, error) {
+	res, err := nrg.controlG.Get()
+
```
```suggestion
```

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+// CurrentPower implements the api.Meter interface
+func (nrg *NRGKickGen2) CurrentPower() (float64, error) {
+	res, err := nrg.valuesG.Get()
+
```
```suggestion
```

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+		return 0, err
+	}
+
+	return res.Powerflow.TotalActivePower, err
```
```suggestion
	return res.Powerflow.TotalActivePower, nil
```

wobu replied:
good catch! thanks

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+		return 0, err
+	}
+
+	return float64(res.Energy.TotalChargedEnergy) * 1e-3, err
```
```suggestion
	return float64(res.Energy.TotalChargedEnergy) * 1e-3, nil
```

andig replied:
This will panic if energy is nil. Either guard or change the struct definition.

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+	return res.Powerflow.L1.Voltage,
+		res.Powerflow.L1.Voltage,
+		res.Powerflow.L1.Voltage,
+		err
```
```suggestion
		nil
```

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+		return 0, err
+	}
+
+	return float64(res.Energy.ChargedEnergy) * 1e-3, err
```
```suggestion
	return float64(res.Energy.ChargedEnergy) * 1e-3, nil
```

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+// Phases1p3p implements the api.PhaseSwitcher interface
+func (nrg *NRGKickGen2) Phases1p3p(phases int) error {
+	res, err := nrg.controlG.Get()
+
```
```suggestion
```

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+
+func (nrg *NRGKickGen2) GetPhases() (int, error) {
+	res, err := nrg.controlG.Get()
+
```
```suggestion
```

andig commented on templates/definition/charger/nrgkick-gen2.yaml:
```diff
@@ -0,0 +1,21 @@
+products:
+  - brand: NRGKick
+    description:
+      generic: Gen2-JSON
```
```suggestion
      generic: Gen2
```

andig commented on templates/definition/charger/nrgkick-gen2.yaml:
```diff
@@ -0,0 +1,21 @@
+      generic: Gen2-JSON
+requirements:
+  description:
+    de: NRGkick Gen2 Ladeeinheit via lokale REST JSON API
```
is the JSON part anyhow relevant for the users?

wobu replied:
there exits also a MODBUS api for the nrgkick gen2, so to avoid confusion i think JSON is important here. In nrgkick gen2 dey call it also JSON or Modbus API

andig replied:
Does it matter for the user? Modbus might be easier to use and a little less overhead. We‘ll add the one or other only.

wobu replied:
to be honest, for the user it wouldn't matter. the JSON API was just easier for me to test and start with.  I will rename everythin to be API agnostic. You'll have to decide if you want to accept the JSON integration or decline it. 

IMHO modbus will have lesser overhead, but i guess it doesn't matter much.

andig commented on README.md:
```diff
@@ -14,7 +14,7 @@ evcc is an extensible EV Charge Controller and home energy management system. Fe
 - simple and clean user interface
 - wide range of supported [chargers](https://docs.evcc.io/docs/devices/chargers):
-  - ABL eMH1, Alfen (Eve), Bender (CC612/613), cFos (PowerBrain), Daheimladen, Ebee (Wallbox), Ensto (Chago Wallbox), [EVSEWifi/ smartWB](https://www.evse-wifi.de), Garo (GLB, GLB+, LS4), go-eCharger, HardyBarth (eCB1, cPH1, cPH2), Heidelberg (Energy Control), Innogy (eBox), Juice (Charger Me), KEBA/BMW, Mennekes (Amedio, Amtron Premium/Xtra, Amtron ChargeConrol), older NRGkicks (before 2022/2023 ...
+  - ABL eMH1, Alfen (Eve), Bender (CC612/613), cFos (PowerBrain), Daheimladen, Ebee (Wallbox), Ensto (Chago Wallbox), [EVSEWifi/ smartWB](https://www.evse-wifi.de), Garo (GLB, GLB+, LS4), go-eCharger, HardyBarth (eCB1, cPH1, cPH2), Heidelberg (Energy Control), Innogy (eBox), Juice (Charger Me), KEBA/BMW, Mennekes (Amedio, Amtron Premium/Xtra, Amtron ChargeConrol), older NRGkicks (before 2022/2023 ...
```
is JSON relevant here?

andig replied:
Rename file to nrg-gen2, in line with the yaml?

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+		return 0, err
+	}
+
+	return res.Powerflow.TotalActivePower, err
```
This will panic if powerflow is nil. Either guard or change the struct definition.

andig commented on charger/nrggen2_json.go:
```diff
@@ -0,0 +1,314 @@
+	enabled  bool
+	controlG provider.Cacheable[gen2.Control]
+	valuesG  provider.Cacheable[gen2.Values]
+	infoG    provider.Cacheable[gen2.Info]
```
This is never used.

wobu committed:
applied requested changes to nrggen2 charger implementation

- renamed to be API implementation agnostic
- removed pointers in types definition
- removed unused code
- fixed error handlings

wobu committed:
removed unused nrgkick gen2 types

wobu committed:
implemented MaxCurrentMillis for nrgkick gen2 charger

wobu commented:
@andig  thanks for your feedback. I do really appreciate. I updated everything as good as i could. The final decision about JSON vs ModBus you will need to decide.

andig commented:
Would prefer modbus if always available. Less code, less overhead.

andig committed:
Update nrggen2.go

andig committed:
Update nrggen2.go

andig commented:
> why do a lot of the modbus implementatins use this decorator to generate code

It's used to add optional methods at runtime. Only needed if capabilities are only discovered when evcc is started (e.h. with or without integrated meter). Is that the case?

> but i had to use a mixture from the github.com/volkszaehler/mbmd/encoding package, where for Uint16 i had to use BigEndian and for >= Uint32 i had to use the methods Int32LswFirst

Endianness is about byte order, but there are different interpretations as to which word order is actually required. If it works its fine whatever you use.

andig commented:
Please open new PR- will close this one. Thank you! Could you also get in contact on Slack?
